### PR TITLE
refactor(nextly): give each write verb one implementation

### PR DESCRIPTION
**Pull request by faisal-rx**
## Summary

Creating, updating and deleting an entry inside a transaction ran through **two implementations apiece**: one for a caller that owns the transaction, and a separate streamlined copy the batch services call per item. Between them they carried 17 clone groups and 699 duplicated lines, and they had already drifted — a comment truncated mid-sentence on one side, a differently worded error message, and a first-publication marker whose rule had to be restated for the batch path after it was found missing there.

Each verb now has **one** implementation that both entry points delegate to:

| shared implementation | replaces | was |
|---|---|---|
| `createEntryWrite` | `createEntryInTransaction` + `createSingleEntryInTransaction` | 470 lines each |
| `updateEntryWrite` | `updateEntryInTransaction` + `updateSingleEntryInTransaction` | 672 / 724 |
| `deleteEntryWrite` | `deleteEntryInTransaction` + `deleteSingleEntryInTransaction` | 318 / 366 |

Plus two extractions every write path repeated: `recordCreateSideEffects` (the post-insert tail — junction rows, version snapshot, outbox event, revalidation intent, all on the caller's transaction) and `readCollectionWriteMeta` (collection + fields + stored hooks + table name).

`collection-mutation-service.ts`: **10,517 → 9,275 lines.**

**No behaviour changes.** Every public method keeps its signature, and `collection-bulk-service.ts` is untouched — all six of its call sites are unchanged.

### The differences are now options, not two bodies

Everything that legitimately differed between each pair is an explicit, named option. If a cell here is wrong, a behaviour moved:

| method | `enforceCollectionAccess` | `rowGate` | `runHooks` | `shapeCallerObject` | `identifyMissingEntry` |
|---|---|---|---|---|---|
| `createEntryInTransaction` | `true` | — | `true` | `true` | — |
| `createSingleEntryInTransaction` | `false` | — | `!skipHooks` | `false` | — |
| `updateEntryInTransaction` | — | `access-service` | `true` | — | `false` |
| `updateSingleEntryInTransaction` | — | `owner-predicate` | `!skipHooks` | — | `true` |
| `deleteEntryInTransaction` | — | `access-service` | `true` | — | `false` |
| `deleteSingleEntryInTransaction` | — | `owner-predicate` | `!skipHooks` | — | `true` |

`enforceCollectionAccess: false` on the batch workers is correct because `collection-bulk-service.ts` hoists that check out of the per-item loop and runs it once per batch, before the transaction opens — verified at all six batch entry points, each of which also pre-resolves the publish transition once.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would change existing behavior)
- [ ] Documentation update
- [x] Refactor / chore (no user-facing change)

## Related issues

None.

## Changeset

- [x] I added a changeset (`.changeset/one-write-path-per-verb.md`, 25 packages, `patch`)
- [x] I selected the correct semver bump (patch / minor / major)

## Test plan

- [x] `pnpm lint`
- [x] `pnpm check-types`
- [x] `pnpm build`
- [x] Manually verified the change

| gate | result |
|---|---|
| `pnpm build` | 19/19 |
| `pnpm --filter nextly check-types` | pass, incl. `tsconfig.tests.json` |
| `pnpm --filter nextly lint` | pass under `--max-warnings 0` |
| `pnpm fallow:audit` | pass — 0 introduced on every dimension |
| pre-push hook | `turbo lint` 22/22, `build` 19/19 |

**Integration, none skipped** (a green run that skipped everything looks identical to one that tested everything, so the skipped count was checked on each leg):

| leg | files / tests |
|---|---|
| Postgres 17 | 16 / 146 |
| SQLite | 12 / 118 |
| MySQL 8 | 4 / 45 |

Covers both `rowGate` modes (`publish-enforcement-owner-only`, `api-key-scope-coverage`), `skipHooks`, working drafts, outbox capture, batch version conflicts, m2m atomicity per dialect, and the first-publication marker on both paths.

**Unit baseline unchanged.** Raw counts differ between checkouts (36 files / 370 failed in the main checkout, 40 / 376 in a worktree) but the four extra files all fail on `Test timed out`, never an assertion, and none touches collections. Control: check out `origin/main`'s copy of the service file and re-run only those four — identical result, 4 files / 7 failed / 147 passed. Per-file counts in the touched area are unchanged: `collection-access` 5, `collection-hooks` 2, `collection-mutation` 4, `related-row-fetch-seam` 3.

### Mutation testing — each option proven load-bearing

A test is evidence only once seen failing for the intended reason. Each break below was applied, observed, and restored:

| break | expected red | observed |
|---|---|---|
| Disable the outbox event in `recordCreateSideEffects` | both callers | **4 tests** in `write-path-events-matrix` — `createEntryInTransaction` ×2 and `createEntries` ×2, `expected 0 to be >= 1`. The 3 update/delete tests stayed green |
| Force `enforceCollectionAccess` false | the create gate | only the new gate test, `expected true to be false`; other 15 green |
| Flip `rowGate` to `access-service` on the batch update | the owner predicate | `keeps the owner predicate for a scoped key on a batch update` |
| Make `skipHooks: true` do nothing | the skip flag | **nothing — see below** |

### Two guards had no real coverage; both are closed here

Found by mutation testing, and both are the same shape — an assertion satisfied by absence:

1. **The collection-level `create` gate on the transaction API.** The only test touching it asserted that `overrideAccess` **bypasses** it, which an entirely absent gate satisfies just as well. Disabling the gate failed nothing. This is the only place that rule is applied to plugins and importers reaching the transaction API. Added its negative twin.
2. **`skipHooks` was never verified to skip anything.** It appears in exactly one test, which asserts a bulk create does not seed field defaults — true whether or not hooks ran. The flag could have been inverted or ignored with every suite green. Added a pair to `code-first-collection-hooks.integration.test.ts`, with a positive control first so the silence means the flag worked rather than that the hook was never registered.

### Reviewer edge cases worth exercising

Not covered by existing tests; each is a real behaviour of this code:

- Roll the caller's transaction back after a successful create → **no row, no junction rows, no version, no outbox event**. All of it is written on `tx`.
- Update/delete a missing id → the transaction API returns `"Entry not found"`; the batch workers return **`"Entry not found: <id>"`** so a failing batch item is identifiable.
- `access.update = owner-only`, batch-update another user's row → the owner predicate is folded into the fetch `WHERE`, so the row never leaves the database and the result is **404, not 403**. The post-fetch comparison is defence-in-depth.
- A super-admin batch-updating another user's owner-only row is **allowed**; the same operation through a **scoped API key owned by a super-admin is denied** — the key is judged on its own grant.
- Failure between the row delete and the outbox insert must **re-throw** and roll back; a failure in an `afterDelete` hook (after the event) must **not**.

## Checklist

- [x] I read [CONTRIBUTING](../CONTRIBUTING.md) (if it exists)
- [x] My commits follow the [Conventional Commits](https://www.conventionalcommits.org/) spec (enforced by commitlint)
- [x] I targeted the `main` branch
- [x] I updated relevant documentation

## Screenshots / recordings

None — server-side only. The integration legs are the equivalent.

## Notes for reviewers

**Read the diff as "the two bodies were identical except for six named things, which are now options"** rather than line by line. The per-pair `diff` that backs that claim: with comments stripped, the create pair differed in **6 places** across 470 lines; update and delete the same way.

**The row gate is deliberately still two shapes selected by an option, not merged.** The batch worker folds an owner predicate into the fetch so a row the caller may not touch never leaves the database; the transaction API fetches by id and asks the access service, which also applies RBAC and route authorization. They disagree today in two ways worth knowing:

- `deleteEntryInTransaction` does not accept `overrideAccess` at all and hard-codes `undefined`, while its batch twin accepts and honours it.
- The update worker carries scoped-API-key context; the delete worker does not.

Reconciling those changes **who may write a row**, so it is left for a change that can be reviewed on its own. This PR makes the asymmetry visible in one place instead of hiding it in two files' worth of near-identical code.

**`shapeCallerObject` preserves an existing quirk.** `createEntryInTransaction` shapes the caller's own `body` object in place — after the call, `tags` has been removed (extracted to the junction), date strings are `Date`s, password fields are hashed. The batch worker copies first. That predates this PR and is unchanged; flagged because it is now visible as an option rather than buried.

**Five suppressions, all recording the same tooling limit.** `fallow` marks a finding `introduced` when a change touches any line inside it, so a refactor inherits every pre-existing finding it moves past — a pure 12-line rename fails the gate on 3 clone groups that exist unchanged on `main`. Each suppression carries a written mechanism reason:

- `.fallowrc.jsonc` → `duplicates.ignoredClones` for one 17-line group shared by `createEntryWrite`/`updateEntryWrite` (validation-gate tail + beforeChange hook calls). They differ in operation, validation mode, `originalData`, and slug re-sanitization; sharing them needs five discriminating parameters to carry seventeen lines of wiring. The key embeds the fingerprint **and** instance count, so any edit to either side makes it report again.
- Four `// fallow-ignore-next-line complexity`. Three are on methods that each **replace two** already-critical findings: the six write methods were 6 findings totalling cyclomatic **189**; they are now 4 totalling **128**. Each comment records the two numbers it replaced. The fourth, `recordCreateSideEffects`, exceeded **CRAP only** (cyclomatic 12, cognitive 9 are both under threshold) — CRAP multiplies complexity by *missing* coverage and the coverage term is `estimated`, not measured; the mutation test above shows it is covered.

**Naming, for anyone new to this file:** `*SingleEntryInTransaction` has nothing to do with Singles. "Single" means *one entry of a batch*. Nextly Singles write through `domains/singles/services/single-entry-service.ts` → `SingleMutationService`, which holds no reference to `CollectionMutationService`; its suites were run and pass (5 files / 40 tests). Renaming these methods was attempted and reverted — it touches lines inside pre-existing clones in `collection-bulk-service.ts` and fails the hygiene gate for that reason alone. Worth doing in the change that dedupes those batch loops.

**Not in this PR:** the interactive `createEntry` / `updateEntry` / `deleteEntry` still hold their own duplication against the transaction paths (12 clone groups / 402 lines). They run hooks, validation and access reads on the pool, *outside* their transaction, while the `*InTransaction` methods run all of it inside the caller's. Making them delegate moves that work into a transaction — changing lock duration and pool behaviour on the write path for all user content. That needs its own change and its own review.

## Summary by CodeRabbit

- **Tests**
  - Added coverage confirming collection hooks run for batch creation unless explicitly skipped.
  - Added coverage verifying collection-level create permissions are enforced when access overrides are disabled.
  - Confirmed denied creation requests do not produce database records.

- **Chores**
  - Documented an intentional duplicate code pattern for static analysis.

### CODE REVIEW

**Comment by faisal-rx** [comment]
@codex @coderabbitai @greptile-apps please review

File: packages/nextly/src/domains/collections/__tests__/code-first-collection-hooks.integration.test.ts [hunk 1]
```diff
@@ -178,4 +179,60 @@ describe("code-first collection hooks (integration)", () => {
...
+
+    // Control: the hooks are registered and the batch path reaches them.
+    const withHooks = await entries.createEntries(
+      { collectionName: "cfhbulk", overrideAccess: true },
+      [{ title: "A" }],
+      { skipHooks: false }
+    );
+    expect(withHooks.failed, JSON.stringify(withHooks)).toBe(0);
+    expect(fired).toContain("beforeChange");
+    expect(fired).toContain("afterChange");
```
[the remark this answers is not in the document]
Comment (faisal-rx) [reply packages/nextly/src/domains/collections/__tests__/code-first-collection-hooks.integration.test.ts]: Agreed, and it matters more here than the Minor badge suggests: this PR merges two copies of each write pipeline into one, so a phase accidentally running twice is precisely the defect it could introduce — and `toContain` is satisfied by exactly that.

Tightened to `expect(fired).toEqual(["beforeChange", "afterChange"])` in `60553a0c5`. It passes, which makes the control prove the stronger property it was always meant to: for one created entry each hook fires once, in order.

Kept the assertion on the line it was on rather than the committable suggestion's two-line form, since the `failed` assertion above it is unrelated.

**Review by mobeenabdullah — APPROVED** [review]
LGTM

**Review by muzzamil-rx — CHANGES_REQUESTED** [review]
This is a strong consolidation — no hard standard violations anywhere: every throw routes through NextlyError, Drizzle-only throughout, no new type-silencing casts, and all access/ownership/404 checks still precede every mutation in all three verbs. The options table was verified against all six call sites and matches exactly. The asks below are about accuracy and residue, not safety.

## Standards

**Ask — orphaned doc comments.** Lines 362–364 of `collection-mutation-service.ts` carry three stacked one-liners (`/** Everything both transaction create entry points accept. */` … delete …) sitting above `interface DeleteEntryWriteParams`; the create/update summaries document nothing, and those two interfaces have no docs of their own. Move each summary onto the interface it describes.

**Ask — a code comment states something the code disproves.** The note above `recordCreateSideEffects` justifies the complexity suppressions with "close to the larger of the two" — but `updateEntryWrite` lands at 59, above **both** predecessors (43/57). The suppression may still be right; the justification as written is false, and these four `fallow-ignore-next-line complexity` directives are the repo's first of their kind in this file — the comment carrying them should say what is true. Relatedly, both suppression comments read as review-response rather than code description.

**Judgement calls, non-blocking:**

- `deleteEntryWrite` (~9000) still computes `collection.schemaDefinition?.fields || collection.fields || []` inline while `readCollectionWriteMeta` now owns exactly that fallback rule for create/update — the one-question-one-implementation residue this PR exists to remove.
- The 404 ternary `options.identifyMissingEntry ? ... : "Entry not found"` repeats 5× (8114, 8476, 8876, 9023, 9067).
- Update and delete carry near-identical ~35-line owner-only post-fetch guards (8121–8170 vs 8893–8937), differing only in verb and update's `isScopedApiKey` carve-out.
- `rowGate: "access-service" | "owner-predicate"` is written twice with no named type, and the gate decision is re-made at 8077/8121 and 8844/8893 — a small `RowGate` type plus a gate-strategy object would collapse both.
- `deleteEntryInTransaction` (8773–8779) re-declares params inline where `updateEntryInTransaction` uses `UpdateEntryWriteParams & { entryId: string }` — pick one shape.

## Spec

Verified correct: one implementation per verb with all six public methods as thin delegators and no third copy; `recordCreateSideEffects` holds the claimed tail on the caller's transaction; line counts exact (10,517 → 9,275); `collection-bulk-service.ts` untouched; signatures preserved member-for-member; the new tests pin the claimed behaviours with real positive controls; and the `.fallowrc.jsonc` clone-group acceptance is properly disclosed in the body's suppressions section. Two asks, both wording in a body reviewers rely on:

- **"two extractions every write path repeated"** overstates: `readCollectionWriteMeta` serves create and update; `deleteEntryWrite` still inlines its own collection/table/hooks read. Delete needing no field list makes that defensible — say so instead of "every write path".
- **"hoists that check … before the transaction opens — verified at all six batch entry points, each of which also pre-resolves the publish transition once"** is half right. All six sites do run `checkCollectionAccess` once per batch, but only the three self-transacting methods run it before the transaction opens — the three `*EntriesInTransaction` variants run it inside the caller's already-open transaction. And `resolveTransitionAuthorization` exists at four of six sites, not each. The `enforceCollectionAccess: false` conclusion still holds; the sentence doesn't.

## Ask

1. Re-home the three orphaned doc comments; make the complexity-suppression note say something true.
2. Correct the two overstated body claims.

The rest is at your discretion; none of it blocks.
